include/circt/Dialect/Verif/VerifOps.td

```diff
@@ -212,4 +212,85 @@ def YieldOp : VerifOp<"yield", [
   ];
 }
 
+//===----------------------------------------------------------------------===//
+// Verification intent ops
+//===----------------------------------------------------------------------===//
+
+/// BMC -- short for bounded model check.
+def BMC : I32EnumAttrCase<"BMC", 0, "bmc">;
+/// KIP -- short for k-Inductive-Proof.
+def KIP : I32EnumAttrCase<"KIP", 0, "induction">;
+
+def FormalTestTypeAttr : I32EnumAttr<"FormalTestType", "formal test type",
+                                [BMC, KIP]> {
+  let cppNamespace = "circt::verif";
+}
+
+def FormalOp : VerifOp<"formal"> {
+  let summary = "defines a formal verification test";
+  let description = [{
+    This operation defines a formal verification test with potentially additional
+    assertions and assumptions. This should be written by declaring symbolic 
+    values that are then connected to a hardware instance. These symbols can then
+    be reasoned about in additional assertions. The region will then be converted 
+    into btor2, SMTLib, or SystemVerilog and ignored during regular SystemVerilog 
+    emission.   
+      
+    Additional information about how the test should be conducted can be passed 
+    using the two supported attributes. These attributes are `k`, which defines 
+    the bounds of the test, and `type` which defines the type of formal test 
+    being done, i.e. Bounded Model Checking (BMC) or k-Induction.
+  }];
+
+  let arguments = (ins OptionalAttr<APIntAttr>:$k,
+                       OptionalAttr<FormalTestTypeAttr>:$test_type,
+                       OptionalAttr<InnerSymAttr>:$inner_sym);
+
+  let regions = (region SizedRegion<1>:$test_region);
+
+  let assemblyFormat = [{
+    attr-dict  (`sym` $inner_sym^)? `(` (`k` `:` $k^)? (`type` `:` $test_type^)? `)` 
+    $test_region
+  }];
+}
+
+//===----------------------------------------------------------------------===//
+// Formal Verification Ops
+//===----------------------------------------------------------------------===//
+
+def SymbolicValOp : VerifOp<"symbolic_val", [Pure]> {
+  let summary = "declare a sybolic value for formal verification";
+  let description = [{
+    This operation declares a symbolic value that can then be used to reason 
+    about hw.instance inputs in a symbolic manner in future assertions and 
+    assumptions. The type result must be explicitly marked.  
+  }];
+  let results = (outs AnyType:$result);
```

Comment:
If I declare an input to a module symbolic with this operation, will it have the same value across clock cycles, can it change any time, can it change according to some clock event?

Reply:
Seconded, and I wonder if a slight tweak to the name would be helpful here - when you talk about a symbolic value in z3, for example, that value can typically only have one assignment. Obviously that setting's a little different since there's no native sense of time, but I wonder if there's another name that could avoid confusion.

Reply:
Yes, this would simply map to an input in btor so it would in theory not have more than one assignment. I think that symbolic_input would be a better name yeah, thanks !

Reply:
Hmm, isn't it one assignment per clock cycle in BTOR internally? One assignment across all clock cycles seems overly restrictive to me. If we go the route of having a single concrete clock (as discussed above), this operation could be specified to hold the same value within a clock cycle of this concrete clock, but (potentially) another value in another cycle?

Reply:
I was under the same impression re: BTOR2 inputs - I think I was a bit unclear in the comment above, but what I was getting at was that an SMT variable can only have one assignment ever, whereas BTOR2 inputs can have different assignments on different clock cycles. I think the specification you're talking about makes the most sense @maerhart - maybe something like an SMT symbolic val (that can only have one assignment ever) would also be useful in some setting, but none spring to my mind